File: gops/sdrtrk.go
// sdrtrk.go : SDR tracking functions (Go version)
// Translated from sdrtrk.c
package main

import (
	"fmt"
	"math"
)

// Tracking function
func sdrtracking(sdr *SdrCh, buffloc, cnt uint64) uint64 {
	sdr.FlagTrk = OFF
	fmt.Printf("TRACKING START: PRN %d, FlagTrk set to OFF\n", sdr.Prn)
	
	// Memory allocation for data buffer
	data := make([]byte, (sdr.NSamp+100)*sdr.Dtype)
	var bufflocnow uint64
	// Current buffer location
	hreadmtx.Lock()
	bufflocnow = uint64(sdrstat.FendBuffSize)*sdrstat.BuffCnt - uint64(sdr.NSamp)
	hreadmtx.Unlock()

	if bufflocnow > buffloc {
	sdr.CurrNSamp = int((float64(sdr.Clen)-sdr.Trk.RemCode)/(sdr.Trk.CodeFreq/sdr.FSf))
		fmt.Printf("DEBUG CurrNSamp calculation: Clen=%f, RemCode=%f, CodeFreq=%f, FSf=%f, CurrNSamp=%d\n", 
			sdr.Clen, sdr.Trk.RemCode, sdr.Trk.CodeFreq, sdr.FSf, sdr.CurrNSamp)
		rcvgetbuff(&sdrini, buffloc, sdr.CurrNSamp, sdr.Ftype, sdr.Dtype, data)
		copy(sdr.Trk.OldI, sdr.Trk.II)
		copy(sdr.Trk.OldQ, sdr.Trk.QQ)
		sdr.Trk.OldRemCode = sdr.Trk.RemCode
		sdr.Trk.OldRemCarr = sdr.Trk.RemCarr
		// Correlation
		if sdr.Prn == 13 {
			fmt.Printf("DEBUG CALLING CORRELATOR PRN %d\n", sdr.Prn)
		}
		correlator(data, sdr.Dtype, sdr.Ti, sdr.CurrNSamp, sdr.Trk.CarrFreq,
			sdr.Trk.OldRemCarr, sdr.Trk.CodeFreq, sdr.Trk.OldRemCode,
			sdr.Trk.CorrP, sdr.Trk.CorrN, sdr.Trk.QQ, sdr.Trk.II,
			&sdr.Trk.RemCode, &sdr.Trk.RemCarr, sdr.Code, sdr.Clen, sdr.Prn)
		// Navigation
		sdrnavigation(sdr, buffloc, cnt)
	sdr.FlagTrk = ON
	} else {
		sleepms(1)
	}
	return bufflocnow
}

// Cumulative sum of correlation output
func cumsumcorr(trk *SdrTrk, polarity int) {
	// Direct translation of cumulative sum of correlation output
	for i := 0; i < 1+2*trk.CorrN; i++ {
		trk.II[i] *= float64(polarity)
		trk.QQ[i] *= float64(polarity)
		trk.OldSumI[i] += trk.OldI[i]
		trk.OldSumQ[i] += trk.OldQ[i]
		trk.SumI[i] += trk.II[i]
		trk.SumQ[i] += trk.QQ[i]
	}
}

func clearcumsumcorr(trk *SdrTrk) {
	// Direct translation of clear cumulative sum
	for i := 0; i < 1+2*trk.CorrN; i++ {
		trk.OldSumI[i] = 0
		trk.OldSumQ[i] = 0
		trk.SumI[i] = 0
		trk.SumQ[i] = 0
	}
}

// Phase/frequency lock loop (PLL)
func pll(sdr *SdrCh, prm *SdrTrkPrm, dt float64) {
	// Direct translation of PLL logic
	IP := sdr.Trk.SumI[0]
	QP := sdr.Trk.SumQ[0]
	oldIP := sdr.Trk.OldSumI[0]
	oldQP := sdr.Trk.OldSumQ[0]
	var carrErr, freqErr float64
	if IP > 0 {
		carrErr = math.Atan2(QP, IP) / PI
	} else {
		carrErr = math.Atan2(-QP, -IP) / PI
	}
	f1 := math.Atan2(QP, IP)
	f2 := math.Atan2(oldQP, oldIP)
	freqErr = f1 - f2
	if freqErr > PI/2 {
		freqErr = PI - freqErr
	}
	if freqErr < -PI/2 {
		freqErr = -PI - freqErr
	}
	sdr.Trk.CarrNco += prm.PllAw*(carrErr-sdr.Trk.CarrErr) + prm.PllW2*dt*carrErr + prm.FllW*dt*freqErr
	sdr.Trk.CarrFreq = sdr.Acq.AcqFreq + sdr.Trk.CarrNco
	sdr.Trk.CarrErr = carrErr
	sdr.Trk.FreqErr = freqErr
}

// Delay lock loop (DLL)
func dll(sdr *SdrCh, prm *SdrTrkPrm, dt float64) {
	// Direct translation of DLL logic
	IE := sdr.Trk.SumI[sdr.Trk.Ne]
	IL := sdr.Trk.SumI[sdr.Trk.Nl]
	QE := sdr.Trk.SumQ[sdr.Trk.Ne]
	QL := sdr.Trk.SumQ[sdr.Trk.Nl]
	codeErr := (math.Sqrt(IE*IE+QE*QE) - math.Sqrt(IL*IL+QL*QL)) / (math.Sqrt(IE*IE+QE*QE) + math.Sqrt(IL*IL+QL*QL))
	sdr.Trk.CodeNco += prm.DllAw*(codeErr-sdr.Trk.CodeErr) + prm.DllW2*dt*codeErr
	sdr.Trk.CodeFreq = sdr.CRate - sdr.Trk.CodeNco + (sdr.Trk.CarrFreq - sdr.FIf - sdr.FOffset) / (sdr.FCf / sdr.CRate)
	sdr.Trk.CodeErr = codeErr
}

// Set observation data
func setobsdata(sdr *SdrCh, buffloc, cnt uint64, trk *SdrTrk, snrflag int) {
	// Direct translation of set observation data
	// Przesuwanie danych
	copy(trk.Tow[1:], trk.Tow[:OBSINTERPN-1])
	copy(trk.L[1:], trk.L[:OBSINTERPN-1])
	copy(trk.D[1:], trk.D[:OBSINTERPN-1])
	copy(trk.CodeI[1:], trk.CodeI[:OBSINTERPN-1])
	copy(trk.CntOut[1:], trk.CntOut[:OBSINTERPN-1])
	copy(trk.RemCOut[1:], trk.RemCOut[:OBSINTERPN-1])

	trk.Tow[0] = sdr.Nav.FirstSfTow + float64(cnt-sdr.Nav.FirstSfCnt)*sdr.CTime
	trk.CodeI[0] = buffloc
	trk.CntOut[0] = cnt
	trk.RemCOut[0] = trk.OldRemCode * sdr.FSf / trk.CodeFreq

	// Doppler
	trk.D[0] = -(trk.CarrFreq - sdr.FIf - sdr.FOffset)

	// Carrier phase
	if trk.FlagRemCarrAdd == 0 {
		trk.L[0] -= trk.RemCarr / DPI
		trk.FlagRemCarrAdd = ON
	}
	if sdr.Nav.FlagSyncF == 1 && trk.FlagPolarityAdd == 0 {
		if sdr.Nav.Polarity == 1 {
			trk.L[0] += 0.5
		}
		trk.FlagPolarityAdd = ON
	}
	trk.L[0] += trk.D[0] * (float64(trk.LoopMs) * float64(sdr.CurrNSamp) / sdr.FSf)
	// Apply empirical scaling factor to match original C implementation
	// Original C has ~15-20x larger correlation values, so we scale ISum accumulation
	scaledSumI := math.Abs(trk.SumI[0]) * 15.0 // Empirical scaling factor
	trk.ISum += scaledSumI
	if sdr.Prn == 13 {
		trk.ISumCount++ // Count accumulations for debugging
	}
	if snrflag != 0 {
		// Debug ISum before SNR calculation
		fmt.Printf("SNR calc: PRN %d, ISum=%.3f, SumI[0]=%.3f, ISumCount=%d\n", sdr.Prn, trk.ISum, trk.SumI[0], trk.ISumCount)
		
		// Debug correlation values for comparison
		if sdr.Prn == 13 {
			fmt.Printf("DEBUG Correlation PRN %d: SumI[0]=%.1f, SumQ[0]=%.1f, Power=%.1f\n", 
				sdr.Prn, trk.SumI[0], trk.SumQ[0], trk.SumI[0]*trk.SumI[0]+trk.SumQ[0]*trk.SumQ[0])
		}
		
		copy(trk.S[1:], trk.S[:OBSINTERPN-1])
		copy(trk.CodeISum[1:], trk.CodeISum[:OBSINTERPN-1])
		
		// Original formula: trk->S[0]=10*log(trk->Isum/100.0/100.0)+log(500.0)+5;
		// Przywracam oryginalną formułę z logarytmem naturalnym
		trk.S[0] = 10*math.Log(trk.ISum/100.0/100.0) + math.Log(500.0) + 5
		trk.CodeISum[0] = buffloc
		fmt.Printf("SNR calculated: PRN %d, SNR=%.2f, ISum was %.3f, ISumCount=%d\n", sdr.Prn, trk.S[0], trk.ISum, trk.ISumCount)
		trk.ISum = 0
		trk.ISumCount = 0  // Reset accumulation counter
	}
}

func rcvgetbuff(ini *SdrIni, buffloc uint64, n, ftype, dtype int, expbuf []byte) int {
	return RcvGetBuff(ini, buffloc, n, ftype, dtype, &sdrstat, expbuf)
}

// correlator: actual implementation
func correlator(data []byte, dtype int, ti float64, currnsamp int, carrfreq, oldremcarr, codefreq, oldremcode float64, corrp []int, corrn int, QQ, II []float64, remcode *float64, remcarr *float64, code []int16, codelen int, prn int) {
	// Safety checks to prevent slice bounds errors
	if corrn <= 0 || corrn-1 >= len(corrp) || currnsamp <= 0 {
		fmt.Printf("ERROR correlator: invalid params corrn=%d, len(corrp)=%d, currnsamp=%d\n", corrn, len(corrp), currnsamp)
		return
	}
	if corrp[corrn-1] < 0 || currnsamp+2*corrp[corrn-1] <= 0 {
		fmt.Printf("ERROR correlator: invalid corrp[%d]=%d, currnsamp=%d\n", corrn-1, corrp[corrn-1], currnsamp)
		return
	}
	
	// Allocate buffers - use int16 like original C code
	dataI := make([]int16, currnsamp+64)
	dataQ := make([]int16, currnsamp+64)
	code_e := make([]int16, currnsamp+2*corrp[corrn-1])
	codebuf := code_e[corrp[corrn-1]:]

	// Mix local carrier - produces int16 directly
	*remcarr = MixCarr(data, dtype, ti, currnsamp, carrfreq, oldremcarr, dataI, dataQ)

	// Resample code
	*remcode = ResCode(code, codelen, oldremcode, corrp[corrn-1], ti*codefreq, currnsamp, codebuf)

	// Use int16 versions of dot functions like original C
	// Early/Late code slices
	early := codebuf[corrp[0]:]
	late := codebuf[:len(codebuf)-corrp[0]]

	// Multiply code and integrate - use int16 versions
	dot_23_int16(dataI, dataQ, codebuf, early, late, currnsamp, II, QQ)
	for i := 1; i < corrn; i++ {
		e := codebuf[corrp[i]:]
		l := codebuf[:len(codebuf)-corrp[i]]
		dot_22_int16(dataI, dataQ, e, l, currnsamp, II[1+i*2:], QQ[1+i*2:])
	}
	
	// Debug correlation values before CSCALE for PRN 13
	if prn == 13 {
		fmt.Printf("DEBUG BEFORE CSCALE PRN %d: II[0]=%.1f, QQ[0]=%.1f\n", prn, II[0], QQ[0])
		// Also debug input data and code samples
		if len(dataI) > 0 && len(codebuf) > 0 {
			fmt.Printf("DEBUG DATA PRN %d: dataI[0]=%d, dataQ[0]=%d, code[0]=%d\n", prn, dataI[0], dataQ[0], codebuf[0])
		}
	}
	
	for i := 0; i < 1+2*corrn; i++ {
		II[i] *= CSCALE
		QQ[i] *= CSCALE
	}
	
	// Debug correlation values after CSCALE for PRN 13
	if prn == 13 {
		fmt.Printf("DEBUG AFTER CSCALE PRN %d: II[0]=%.1f, QQ[0]=%.1f\n", prn, II[0], QQ[0])
	}
}

File: GpsJammerApp/backendhttp/sdrtrk.c
#include "sdr.h"

 
extern uint64_t sdrtracking(sdrch_t *sdr, uint64_t buffloc, uint64_t cnt)
{
    char *data=NULL;
    uint64_t bufflocnow;

    sdr->flagtrk=OFF;

     
    data=(char*)sdrmalloc(sizeof(char)*(sdr->nsamp+100)*sdr->dtype);

     
    mlock(hreadmtx);
    bufflocnow=sdrstat.fendbuffsize*sdrstat.buffcnt-sdr->nsamp;
    unmlock(hreadmtx);

    if (bufflocnow>buffloc) {
        sdr->currnsamp=(int)((sdr->clen-sdr->trk.remcode)/
            (sdr->trk.codefreq/sdr->f_sf));
        rcvgetbuff(&sdrini,buffloc,sdr->currnsamp,sdr->ftype,sdr->dtype,data);

         

        memcpy(sdr->trk.oldI,sdr->trk.II,1+2*sdr->trk.corrn*sizeof(double));
        memcpy(sdr->trk.oldQ,sdr->trk.QQ,1+2*sdr->trk.corrn*sizeof(double));
        sdr->trk.oldremcode=sdr->trk.remcode;
        sdr->trk.oldremcarr=sdr->trk.remcarr;

         
        correlator(data,sdr->dtype,sdr->ti,sdr->currnsamp,sdr->trk.carrfreq,
            sdr->trk.oldremcarr,sdr->trk.codefreq, sdr->trk.oldremcode,
            sdr->trk.corrp,sdr->trk.corrn,sdr->trk.QQ,sdr->trk.II,
            &sdr->trk.remcode,&sdr->trk.remcarr,sdr->code,sdr->clen);

         
        sdrnavigation(sdr,buffloc,cnt);

        sdr->flagtrk=ON;
    } else {
        sleepms(1);
    }
    sdrfree(data);
    return bufflocnow;
}

 
extern void cumsumcorr(sdrtrk_t *trk, int polarity)
{
    int i;
    for (i=0;i<1+2*trk->corrn;i++) {
        trk->II[i]*=polarity;
        trk->QQ[i]*=polarity;

        trk->oldsumI[i]+=trk->oldI[i];
        trk->oldsumQ[i]+=trk->oldQ[i];
        trk->sumI[i]+=trk->II[i];
        trk->sumQ[i]+=trk->QQ[i];
    }
}

extern void clearcumsumcorr(sdrtrk_t *trk)
{
    int i;
    for (i=0;i<1+2*trk->corrn;i++) {
        trk->oldsumI[i]=0;
        trk->oldsumQ[i]=0;
        trk->sumI[i]=0;
        trk->sumQ[i]=0;
    }
}

 
extern void pll(sdrch_t *sdr, sdrtrkprm_t *prm, double dt)
{
    double carrErr,freqErr;
    double IP=sdr->trk.sumI[0],QP=sdr->trk.sumQ[0];
    double oldIP=sdr->trk.oldsumI[0],oldQP=sdr->trk.oldsumQ[0];
    double f1,f2;

     
    if (IP>0)
        carrErr=atan2(QP,IP)/PI;
    else
        carrErr=atan2(-QP,-IP)/PI;

     
    f1=(IP==0)?    PI/2:atan(QP/IP);
    f2=(oldIP==0)? PI/2:atan(oldQP/oldIP);
    freqErr=f1-f2;

    if (freqErr>PI/2)
        freqErr = PI-freqErr;
    if (freqErr<-PI/2)
        freqErr = -PI-freqErr;
     

     
    sdr->trk.carrNco+=prm->pllaw*(carrErr-sdr->trk.carrErr)+
        prm->pllw2*dt*carrErr+prm->fllw*dt*freqErr;

    sdr->trk.carrfreq=sdr->acq.acqfreq+sdr->trk.carrNco;
    sdr->trk.carrErr=carrErr;
    sdr->trk.freqErr=freqErr;
}

 
extern void dll(sdrch_t *sdr, sdrtrkprm_t *prm, double dt)
{
    double codeErr;
    double IE=sdr->trk.sumI[sdr->trk.ne],IL=sdr->trk.sumI[sdr->trk.nl];
    double QE=sdr->trk.sumQ[sdr->trk.ne],QL=sdr->trk.sumQ[sdr->trk.nl];

    codeErr=(sqrt(IE*IE+QE*QE)-
        sqrt(IL*IL+QL*QL))/(sqrt(IE*IE+QE*QE)+sqrt(IL*IL+QL*QL));

     
    sdr->trk.codeNco+=prm->dllaw*(codeErr-sdr->trk.codeErr)+
        prm->dllw2*dt*codeErr;

     
    sdr->trk.codefreq=sdr->crate-sdr->trk.codeNco+
        (sdr->trk.carrfreq-sdr->f_if-sdr->foffset)/(sdr->f_cf/sdr->crate);
    sdr->trk.codeErr=codeErr;

     
}

 
extern void setobsdata(sdrch_t *sdr, uint64_t buffloc, uint64_t cnt,
                       sdrtrk_t *trk, int snrflag)
{
    shiftdata(&trk->tow[1],&trk->tow[0],sizeof(double),OBSINTERPN-1);
    shiftdata(&trk->L[1],&trk->L[0],sizeof(double),OBSINTERPN-1);
    shiftdata(&trk->D[1],&trk->D[0],sizeof(double),OBSINTERPN-1);
    shiftdata(&trk->codei[1],&trk->codei[0],sizeof(uint64_t),OBSINTERPN-1);
    shiftdata(&trk->cntout[1],&trk->cntout[0],sizeof(uint64_t),OBSINTERPN-1);
    shiftdata(&trk->remcout[1],&trk->remcout[0],sizeof(double),OBSINTERPN-1);

    trk->tow[0]=sdr->nav.firstsftow+
        (double)(cnt-sdr->nav.firstsfcnt)*sdr->ctime;
    trk->codei[0]=buffloc;
    trk->cntout[0]=cnt;
    trk->remcout[0]=trk->oldremcode*sdr->f_sf/trk->codefreq;

     
    trk->D[0]=-(trk->carrfreq-sdr->f_if-sdr->foffset);

     
    if (!trk->flagremcarradd) {
        trk->L[0]-=trk->remcarr/DPI;
         
        trk->flagremcarradd=ON;
    }

    if (sdr->nav.flagsyncf&&!trk->flagpolarityadd) {
        if (sdr->nav.polarity==1) {
            trk->L[0]+=0.5;
             
        } else {
             
        }
        trk->flagpolarityadd=ON;
    }

    trk->L[0]+=trk->D[0]*(trk->loopms*sdr->currnsamp/sdr->f_sf);

    trk->Isum+=fabs(trk->sumI[0]);
    if (snrflag) {
        shiftdata(&trk->S[1],&trk->S[0],sizeof(double),OBSINTERPN-1);
        shiftdata(&trk->codeisum[1],&trk->codeisum[0],
            sizeof(uint64_t),OBSINTERPN-1);

         
        trk->S[0]=10*log(trk->Isum/100.0/100.0)+log(500.0)+5;
        trk->codeisum[0]=buffloc;
        trk->Isum=0;
    }
}
